File: code-examples/c/monitoring/apm-subscribe.c
/* Include necessary MongoDB C libraries */
#include <mongoc.h>

/* Function declaration for event handler*/
static void
event_handler(const mongoc_apm_command_succeeded_t *event);

int
main (int argc, char *argv[])
{
  mongoc_apm_callbacks_t *callbacks;
  mongoc_client_t *client;
  mongoc_uri_t *uri;

  mongoc_init ();

  // Replace the following with your MongoDB deployment's connection string
  uri = mongoc_uri_new ("mongodb+srv://<clusterUrl>/?replicaSet=rs&writeConcern=majority");

  client = mongoc_client_new_from_uri (uri);

  // Add the event handler to monitor a specified command
  callbacks = mongoc_apm_callbacks_new ();
  // Replace <event name> with the name of the event you are subscribing to
  mongoc_apm_set_command_succeeded_cb (callbacks, event_handler);

  mongoc_client_set_apm_callbacks (client, callbacks, NULL);

  // Ping the "admin" database to verify connection
  bson_t *command;
  bson_t reply;
  bson_error_t error;
  command = BCON_NEW ("ping", BCON_INT32 (1));
  if (mongoc_client_command_simple(client, "admin", command, NULL, &reply, &error)) {
    printf ("Connected successfully\n");
  } else {
    fprintf (stderr, "%s\n", error.message);
  }

  mongoc_client_destroy (client);
  mongoc_uri_destroy (uri);
  mongoc_cleanup ();

  return EXIT_SUCCESS;
}

// Event handler function
static void
event_handler(const mongoc_apm_command_succeeded_t *event)
{
  char *event_json;
  // Log event data when it's emitted
  event_json = bson_as_canonical_extended_json (
      mongoc_apm_command_succeeded_get_reply (event), NULL);
  printf("Received event: %s\n", event_json);
  bson_free (event_json);
}

File: code-examples/php/monitoring/apm-subscribe.php
<?php

require_once './vendor/autoload.php';

// Replace the following with your MongoDB deployment's connection string
$uri = "mongodb+srv://<clusterUrl>/?replicaSet=rs&writeConcern=majority";

$client = new MongoDB\Client($uri);

// Replace <event name> with the name of the event you are subscribing to
$eventName = "<event name>";

$client->getManager()->addSubscriber(new MongoDB\Driver\Monitoring\CommandSubscriber(
    function(MongoDB\Driver\Monitoring\CommandStartedEvent $event) use ($eventName) {
        if ($eventName === $event->getCommandName()) {
            echo "received " . $eventName . ": " . json_encode($event->getCommand()) . "\n";
        }
    }
));

try {
    // Establish and verify connection to the "admin" database
    $command = new MongoDB\Driver\Command(array( "ping" => 1 ));
    $client->getManager()->executeCommand("admin", $command);
    echo "Connected successfully\n";
} finally {
    // Close the database connection on completion or error
    $client = null;
}

?>
